File: 2.Add Two Numbers/add_two_numbers.go
package main

import "fmt"

/**
* Definition for singly-linked list.
* type ListNode struct {
	*     Val int
	*     Next *ListNode
	* }
*/

type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	head := &ListNode{Val: 0}
	n1, n2, carry, current := 0, 0, 0, head
	for l1 != nil || l2 != nil || carry != 0 {
		if l1 != nil {
			n1 = l1.Val
			l1 = l1.Next
		} else {
			n1 = 0
		}

		if l2 != nil {
			n2 = l2.Val
			l2 = l2.Next
		} else {
			n2 = 0
		}

		current.Next = &ListNode{Val: (n1 + n2 + carry) % 10}
		current = current.Next
		carry = (n1 + n2 + carry) / 10
	}

	return head.Next

}

func main() {
	l1 := &ListNode{Val: 1}
	l1.Next = &ListNode{Val: 7}
	l1.Next.Next = &ListNode{Val: 9}

	l2 := &ListNode{Val: 8}
	l2.Next = &ListNode{Val: 9}
	l2.Next.Next = &ListNode{Val: 3}

	result := addTwoNumbers(l1, l2)

	fmt.Println(result.Val)
	p := result.Next
	for p != nil {
		fmt.Print(p.Val)
		if p.Next == nil {
			break
		} else {
			p = p.Next
		}
	}
}

File: 2. Add Two Numbers/add_two_numbers.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def addTwoNumber(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        head = ListNode(val=0)
        n1 = 0
        n2 = 0
        carry = 0
        current = head

        while l1 is not None or l2 is not None or carry != 0:
            if l1 is not None:
                n1 = l1.val
                l1 = l1.next
            else:
                n1 = 0

            if l2 is not None:
                n2 = l2.val
                l2 = l2.next
            else:
                n2 = 0

            current.next = ListNode(val=(n1+n2+carry) % 10)
            current = current.next
            carry = (n1+n2+carry)//10

        return head.next


def main():
    l11 = ListNode(val=8)
    l12 = ListNode(val=3, next=l11)
    l13 = ListNode(val=6, next=l12)

    l21 = ListNode(val=7)
    l22 = ListNode(val=1, next=l21)
    l23 = ListNode(val=9, next=l22)

    s = Solution()
    result = s.addTwoNumber(l13, l23)
    print(result.val)
    p = result.next
    while p is not None:
        print(p.val)
        if p.next is None:
            break
        else:
            p = p.next


if __name__ == "__main__":
    main()
